### PR TITLE
issue #24644 add copy constructor with partial value override

### PR DESCRIPTION
**Pull request by pcornelissen**
I have created a small pull request that provides the necessary functionality that I wrote about in my feature request #24644. I have added isUTC copying as well.
I have already submitted the CLA for stuff I did for angularDart.

### CODE REVIEW

**Comment by mit-mit** [comment]
@pcornelissen, can you take a look at the questions from @whesse, please?

File: sdk/lib/core/date_time.dart [hunk 1]
```diff
@@ -184,6 +184,41 @@ class DateTime implements Comparable {
...
+  DateTime.from(
+      DateTime src, {
+      int year : null,
+      int month : null,
+      int day : null,
+      int hour : null,
+      int minute : null,
+      int second : null,
+      int millisecond : null,
+      bool isUtc:null
```
Comment (lrhn) [inline sdk/lib/core/date_time.dart]: spaces around `:`.

File: sdk/lib/core/date_time.dart [hunk 2]
```diff
@@ -184,6 +184,41 @@ class DateTime implements Comparable {
...
+      }):this._internal(
```
Comment (lrhn) [inline sdk/lib/core/date_time.dart]: spaces around this `:` too.

File: sdk/lib/core/date_time.dart [hunk 3]
```diff
@@ -184,6 +184,41 @@ class DateTime implements Comparable {
...
+        year == null ? src.year : year,
```
Comment (lrhn) [inline sdk/lib/core/date_time.dart]: You can use the `??~ operator:  `year = year ?? src.year,`.

File: tests/corelib/date_time10_test.dart [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  Expect.equals(1999, dt1.year);
+  Expect.equals(1, dt1.month);
+  Expect.equals(2, dt1.day);
+  Expect.equals(20, dt1.hour);
+  Expect.equals(10, dt1.minute);
+  Expect.equals(5, dt1.second);
+  Expect.equals(432, dt1.millisecond);
+  Expect.equals(false, dt1.isUtc);
+
+}
```
Comment (lrhn) [inline tests/corelib/date_time10_test.dart]: More tests. Try setting each field individually, use too-large and too-small (negative or zero) values.
I would probably make a test helper function:
```
   test(date, {year, ... , milliseconds,isUtc}) {
     var expected = new DateTime(year ?? date.year, ... , isUtc ? date.isUtc);
     var actual = date.replace(year: year, ... , isUtc; isUtc);
     Expect.equals(expected, actual);
  }
```
That allows you to write may tests in little space.

File: sdk/lib/core/date_time.dart [hunk 5]
```diff
@@ -184,6 +184,41 @@ class DateTime implements Comparable {
...
+   *
+   *     DateTime annularEclipse = new DateTime(2014, DateTime.APRIL, 29, 6, 4);
+   *
+   * Now you need a second DateTime on another Day, but the rest should stay the same
+   *
+   *     DateTime second = new DateTime.from(annularEclipse, month: DateTime.SEPTEMBER, day: 20);
+   */
+  DateTime.from(
+      DateTime src, {
+      int year : null,
```
Comment (lrhn) [inline sdk/lib/core/date_time.dart]: I think the normal style is to put no space before the ':' and one space after:
`int year: null,`

File: tests/corelib/date_time10_test.dart (hunk 4 above)
Comment (pcornelissen) [inline tests/corelib/date_time10_test.dart]: Well, my plan was not to rewrite the tests for datetime, just some functionality that makes the life much easier when you really have to work with the api. Feel free to improve it ;-)

File: sdk/lib/core/date_time.dart (hunk 5 above)
Comment (kevmoo) [inline sdk/lib/core/date_time.dart]: No need to include default values of null, actually.

**Comment by mit-mit** [comment]
Closing this as no response has been received in the above questions. @pcornelissen you are most welcome to re-open it should you have a chance to address the questions.

**Comment by pcornelissen** [comment]
Well I just wanted to provide some code to help you improve the developer experience, if you don't want it ok. Feel free to use this or do something similar. I don't care ;-)

**Comment by lrhn** [comment]
The idea is good, but instead of making it a constructor, I think it fits
better as a method on the source DateTime. We have a CL for this (
https://codereview.chromium.org/1472803003/), but it's currently blocked on
naming.

**Pull request closed without merging**